pyodide-build/pyodide_build/out_of_tree/pypi.py

```diff
@@ -0,0 +1,386 @@
+            os.dup2(copied.fileno(), stream_fd)  # $ exec >&copied
+            to = None
+
+
+# cache of built sdists by url source to wheel file path
+BUILD_CACHE: dict[str, dict[str, Any]] = {}
+
+
+def get_cached_built_wheel(url):
+    parsed_url = urlparse(url)
+    gz_name = Path(parsed_url.path).name
+    if url in BUILD_CACHE:
+        return BUILD_CACHE[url]["path"]
+    else:
+        cache_entry = {}
+        build_dir = tempfile.TemporaryDirectory()
+        cache_entry["build_dir"] = build_dir
+        os.chdir(build_dir.name)
+        with tempfile.NamedTemporaryFile(suffix=".tar.gz", delete=False) as f:
+            data = requests.get(url).content
+            f.write(data)
+            f.close()
+            shutil.unpack_archive(f.name, build_dir.name)
+            os.unlink(f.name)
+            files = list(os.listdir(build_dir.name))
+            if len(files) == 1:
+                os.chdir(os.path.join(build_dir.name, files[0]))
+            else:
+                os.chdir(build_dir.name)
+        print(f"Building wheel for {gz_name}: ", end="")
+        with tempfile.TemporaryFile(mode="w+") as f:
+            try:
+                with stream_redirected(to=f, stream=sys.stdout):
+                    with stream_redirected(to=f, stream=sys.stderr):
+                        wheel_path = build.run(
+                            PyPIProvider.BUILD_EXPORTS, PyPIProvider.BUILD_FLAGS
+                        )
+            except BaseException as e:
+                print(" Failed\n Error is:")
+                f.seek(0)
+                sys.stdout.write(f.read())
+                raise e
+
+        OKGREEN = "\033[92m"
+        ENDC = "\033[0m"
+        print(OKGREEN, "Success", ENDC)
+
+        cache_entry["path"] = wheel_path
+        BUILD_CACHE[url] = cache_entry
+        return wheel_path
+
+
+class Candidate:
+    def __init__(self, name, version, url=None, extras=None):
+        self.name = canonicalize_name(name)
+        self.version = version
+        self.url = url
+        self.extras = extras
+
+        self._metadata = None
+        self._dependencies = None
+
+    def __repr__(self):
+        if not self.extras:
+            return f"<{self.name}=={self.version}>"
+        return f"<{self.name}[{','.join(self.extras)}]=={self.version}>"
+
+    @property
+    def metadata(self):
+        if self._metadata is None:
+            self._metadata = get_metadata_for_wheel(self.url)
+        return self._metadata
+
+    @property
+    def requires_python(self):
+        return self.metadata.get("Requires-Python")
+
+    def _get_dependencies(self):
+        deps = self.metadata.get_all("Requires-Dist", [])
+        extras = self.extras if self.extras else [""]
+
+        for d in deps:
+            r = Requirement(d)
+            if r.marker is None:
+                yield r
+            else:
+                for e in extras:
+                    if r.marker.evaluate({"extra": e}):
+                        yield r
```

Comment:
Probably this should be:
```suggestion
                        yield r
                        break
```
so that the requirement isn't yielded multiple times?